storage/umass/umass.c

```diff
@@ -324,173 +627,273 @@ static void umass_msgthr(void *arg)
+	int err;
+	oid_t roid;
+
+	do {
+		err = umass_mountFromDev(dev, LIBEXT2_NAME, &roid);
+		if (err < 0) {
+			break;
+		}
+		err = portRegister(roid.port, "/", &roid);
+		if (err < 0) {
+			break;
+		}
+		err = EOK;
+	} while (0);
+
+	return err;
+#else
+	return -ENOSYS;
+#endif
+}
+
+
+static int umass_handleInsertion(usb_driver_t *drv, usb_devinfo_t *insertion)
 {
+	int err;
 	umass_dev_t *dev;
 	oid_t oid;
 
+	mutexLock(umass_common.lock);
+
 	if ((dev = umass_devAlloc()) == NULL) {
 		fprintf(stderr, "umass: devAlloc failed\n");
 		return -ENOMEM;
 	}
 
+	dev->drv = drv;
 	dev->instance = *insertion;
-	if ((dev->pipeCtrl = usb_open(insertion, usb_transfer_control, 0)) < 0) {
+	dev->pipeCtrl = usb_open(drv, insertion, usb_transfer_control, 0);
+	if (dev->pipeCtrl < 0) {
 		free(dev);
 		fprintf(stderr, "umass: usb_open failed\n");
 		return -EINVAL;
 	}
 
-	if (usb_setConfiguration(dev->pipeCtrl, 1) != 0) {
+	err = usb_setConfiguration(drv, dev->pipeCtrl, 1);
+	if (err != 0) {
 		free(dev);
 		fprintf(stderr, "umass: setConfiguration failed\n");
 		return -EINVAL;
 	}
 
-	if ((dev->pipeIn = usb_open(insertion, usb_transfer_bulk, usb_dir_in)) < 0) {
+	dev->pipeIn = usb_open(drv, insertion, usb_transfer_bulk, usb_dir_in);
+	if (dev->pipeIn < 0) {
 		fprintf(stderr, "umass: pipe open failed \n");
 		free(dev);
 		return -EINVAL;
 	}
 
-	if ((dev->pipeOut = usb_open(insertion, usb_transfer_bulk, usb_dir_out)) < 0) {
+	dev->pipeOut = usb_open(drv, insertion, usb_transfer_bulk, usb_dir_out);
+	if (dev->pipeOut < 0) {
 		fprintf(stderr, "umass: pipe open failed\n");
 		free(dev);
 		return -EINVAL;
 	}
 	dev->tag = 0;
 
-	if (umass_check(dev)) {
+	err = _umass_check(dev);
+	if (err != 0) {
 		fprintf(stderr, "umass: umass_check failed\n");
 		free(dev);
 		return -EINVAL;
 	}
 
 	oid.port = umass_common.msgport;
 	oid.id = dev->fileId;
-	if (create_dev(&oid, dev->path) != 0) {
+	err = create_dev(&oid, dev->path);
+	if (err != 0) {
 		free(dev);
 		fprintf(stderr, "usb: Can't create dev!\n");
 		return -EINVAL;
 	}
 
-	LIST_ADD(&umass_common.devices, dev);
-	fprintf(stderr, "umass: New USB Mass Storage device: %s sectors: %d\n", dev->path, dev->partSize);
+	printf("umass: New USB Mass Storage device: %s sectors: %d\n", dev->path, dev->part.sectors);
+
+	mutexUnlock(umass_common.lock);
+
+	if (umass_common.mount_root) {
+		err = umass_mountRoot(dev);
+		if (err < 0) {
+			fprintf(stderr, "umass: failed to mount root partition\n");
+			return err;
+		}
+		umass_common.mount_root = false; /* don't try to mount root again */
+	}
 
 	return 0;
 }
 
 
-static int umass_handleDeletion(usb_deletion_t *del)
+static int umass_handleDeletion(usb_driver_t *drv, usb_deletion_t *del)
 {
-	umass_dev_t *next, *dev = umass_common.devices;
-	int cont = 1;
+	umass_dev_t *dev;
+	rbnode_t *node;
 
-	if (dev == NULL)
-		return 0;
+	mutexLock(umass_common.lock);
+
+	for (node = lib_rbMinimum(umass_common.devices.root); node != NULL; node = lib_rbNext(node)) {
```

Comment:
lib_rbNext(node) after node has been freed (node is a part of dev) is wrong in two ways:
- node is no longer a part of the tree, so it became invalid as an iterator,
- use after free.

You need to reset the iterator to a valid one before removing current one